include/irr/asset/ITexturePacker.h

```diff
@@ -0,0 +1,483 @@
+
+    uint32_t maxAllocatableTextureSz() const { return m_pgSzxy<<m_pageTable->getCreationParameters().mipLevels; }
+
+    uint32_t physAddrTexLayerSz() const
+    {
+        return m_tilesPerDim * (2u*TILE_PADDING+m_pgSzxy);
+    }
+    //! @returns texel-wise offset of physical page
+    core::vector3du32_SIMD pageCoords(SPhysPgOffset _txoffset, uint32_t _pgSz) const
+    {
+        core::vector3du32_SIMD coords(physPgOffset_x(_txoffset), physPgOffset_y(_txoffset), 0u);
+        coords *= (_pgSz + 2u*TILE_PADDING);
+        coords += TILE_PADDING;
+        coords.z = physPgOffset_layer(_txoffset);
+        return coords;
+    }
+    struct SMiptailPacker
+    {
+        struct rect
+        {
+            int x, y, mx, my;
+        };
+        static bool computeMiptailOffsets(rect* res, int log2SIZE, int padding=9);
+    };
+
+    ICPUTexturePacker(E_FORMAT_CLASS _fclass, E_FORMAT _format, uint32_t _pgTabSzxy, uint32_t _maxMipLevels, uint32_t _pgSzxy = 256u, uint32_t _tilesPerDim = 32u, uint32_t _numLayers = 4u, uint32_t _tilePad = 9u/*max_aniso/2+1*/) :
+        ITexturePacker(_pgTabSzxy, _pgSzxy, _tilesPerDim), TILE_PADDING(_tilePad),
+        m_physPgAddrAlctr_reservedSpc(core::make_refctd_dynamic_array<decltype(m_physPgAddrAlctr_reservedSpc)>(phys_pg_addr_alctr_t::reserved_size(1u, _numLayers*_tilesPerDim*_tilesPerDim, 1u))),
+        m_physPgAddrAlctr(m_physPgAddrAlctr_reservedSpc->data(), 0u, 0u, 1u, _numLayers*_tilesPerDim*_tilesPerDim, 1u)
+    {
+        assert(getFormatClass(_format)==_fclass);
+        {
+            const uint32_t SZ = physAddrTexLayerSz();
+
+            ICPUImage::SCreationParams params;
+            params.extent = { SZ,SZ,1u };
+            params.format = _format;
+            params.arrayLayers = _numLayers;
+            params.mipLevels = 1u;
+            params.type = IImage::ET_2D;
+            params.samples = IImage::ESCF_1_BIT;
+            params.flags = static_cast<IImage::E_CREATE_FLAGS>(0);
+
+            m_physAddrTex = ICPUImage::create(std::move(params));
+            auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<IImage::SBufferCopy>>(1ull);
+            auto& region = regions->front();
+            region.imageSubresource.mipLevel = 0u;
+            region.imageSubresource.baseArrayLayer = 0u;
+            region.imageSubresource.layerCount = _numLayers;
+            region.bufferOffset = 0u;
+            region.bufferRowLength = SZ;
+            region.bufferImageHeight = 0u; //tightly packed
+            region.imageOffset = { 0u, 0u, 0u };
+            region.imageExtent = { SZ,SZ,1u };
+            auto buffer = core::make_smart_refctd_ptr<ICPUBuffer>(getTexelOrBlockBytesize(_format) * SZ * SZ * _numLayers);
+            m_physAddrTex->setBufferAndRegions(std::move(buffer), regions);
+        }
+        {
+            ICPUImage::SCreationParams params;
+            params.arrayLayers = 1u;
```

Comment:
because of IEEE754 precision issues once you start having virtual UV coordinates for textures that are > 32k in size (256 pages) we should support layered Page Tables and UV offsets